### PR TITLE
Fix preference not taken into account for `odo version`

### LINKED ISSUE #6417: Timeout not taken into account during `odo version`
/kind bug

When running `odo version`, the Timeout preference is not considered, as it should be.

### PR DESCRIPTION
**Pull request by feloy**
**What type of PR is this:**

/kind bug

**What does this PR do / why we need it:**

This PR adds the PREFERNECE dependency to `odo version`, and creates a CommandGroup abstraction to avoid this error occurs again.

**Which issue(s) this PR fixes:**

`odo version` can use the Timeout value defined in the odo preferences, but this value were never taken into account, because the PREFERENCE dependency were overwritten when setting the command group.

Fixes #6417

**PR acceptance criteria:**

- [ ] Unit test 

- [ ] Integration test 

- [ ] Documentation 

**How to test changes / Special notes to the reviewer:**

### CODE REVIEW

**Review by rm3l — COMMENTED** [review]
Good catch - thanks!

I'm wondering in this case if we shouldn't make the preference client a mandatory dependency, and explicitly fail if this is not the case.
I can see that `preferenceInfo#GetTimeout()` already falls back to `DefaultTimeout` if there is no timeout setting defined. So we could be more strict by expecting the Preference client dependency to be there in `VersionOptions#Complete`.
To be clear, I am just thinking of the following diff:
```diff
diff --git a/pkg/odo/cli/version/version.go b/pkg/odo/cli/version/version.go
index 180c8b58a..e31226808 100644
--- a/pkg/odo/cli/version/version.go
+++ b/pkg/odo/cli/version/version.go
@@ -5,13 +5,11 @@ import (
        "fmt"
        "os"
        "strings"
-       "time"
 
        "github.com/redhat-developer/odo/pkg/kclient"
        "github.com/redhat-developer/odo/pkg/odo/cmdline"
        "github.com/redhat-developer/odo/pkg/odo/genericclioptions"
        "github.com/redhat-developer/odo/pkg/odo/genericclioptions/clientset"
-       "github.com/redhat-developer/odo/pkg/preference"
        odoversion "github.com/redhat-developer/odo/pkg/version"
 
        "github.com/spf13/cobra"
@@ -63,16 +61,7 @@ func (o *VersionOptions) Complete(ctx context.Context, cmdline cmdline.Cmdline,
                client, err := kclient.New()
 
                if err == nil {
-                       // checking the value of timeout in preference
-                       var timeout time.Duration
-                       if o.clientset.PreferenceClient != nil {
-                               timeout = o.clientset.PreferenceClient.GetTimeout()
-                       } else {
-                               // the default timeout will be used
-                               // when the value is not readable from preference
-                               timeout = preference.DefaultTimeout
-                       }
-                       o.serverInfo, err = client.GetServerVersion(timeout)
+                       o.serverInfo, err = client.GetServerVersion(o.clientset.PreferenceClient.GetTimeout())
                        if err != nil {
                                klog.V(4).Info("unable to fetch the server version: ", err)
                        }
```

At least, this would explicitly panic if the Preference Client is not there and allow to catch this issue if it occurs again, IMO.

What do you think?

**Comment by feloy** [comment]
> Good catch - thanks!
> 
> I'm wondering in this case if we shouldn't make the preference client a mandatory dependency, and explicitly fail if this is not the case. I can see that `preferenceInfo#GetTimeout()` already falls back to `DefaultTimeout` if there is no timeout setting defined. So we could be more strict by expecting the Preference client dependency to be there in `VersionOptions#Complete`. To be clear, I am just thinking of the following diff:
> 
> ```diff
> diff --git a/pkg/odo/cli/version/version.go b/pkg/odo/cli/version/version.go
> index 180c8b58a..e31226808 100644
> --- a/pkg/odo/cli/version/version.go
> +++ b/pkg/odo/cli/version/version.go
> @@ -5,13 +5,11 @@ import (
>         "fmt"
>         "os"
>         "strings"
> -       "time"
>  
>         "github.com/redhat-developer/odo/pkg/kclient"
>         "github.com/redhat-developer/odo/pkg/odo/cmdline"
>         "github.com/redhat-developer/odo/pkg/odo/genericclioptions"
>         "github.com/redhat-developer/odo/pkg/odo/genericclioptions/clientset"
> -       "github.com/redhat-developer/odo/pkg/preference"
>         odoversion "github.com/redhat-developer/odo/pkg/version"
>  
>         "github.com/spf13/cobra"
> @@ -63,16 +61,7 @@ func (o *VersionOptions) Complete(ctx context.Context, cmdline cmdline.Cmdline,
>                 client, err := kclient.New()
>  
>                 if err == nil {
> -                       // checking the value of timeout in preference
> -                       var timeout time.Duration
> -                       if o.clientset.PreferenceClient != nil {
> -                               timeout = o.clientset.PreferenceClient.GetTimeout()
> -                       } else {
> -                               // the default timeout will be used
> -                               // when the value is not readable from preference
> -                               timeout = preference.DefaultTimeout
> -                       }
> -                       o.serverInfo, err = client.GetServerVersion(timeout)
> +                       o.serverInfo, err = client.GetServerVersion(o.clientset.PreferenceClient.GetTimeout())
>                         if err != nil {
>                                 klog.V(4).Info("unable to fetch the server version: ", err)
>                         }
> ```
> 
> At least, this would explicitly panic if the Preference Client is not there and allow to catch this issue if it occurs again, IMO.
> 
> What do you think?

You are right, I'm also annoyed by this test on the client, which was hiding the error.

I'll apply your patch.
Thanks

File: cmd/odo/help_test.go [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+	utilityCommands = `Utility Commands:
+  analyze      Detect devfile to use based on files present in current directory
+  completion   Add odo completion support to your development environment
+  preference   Modifies preference settings (add, remove, set, unset, view)
+  version      Print the client version information
+
+`
+)
+
+func TestOdoHelp(t *testing.T) {
```
Comment (rm3l) [inline cmd/odo/help_test.go:64]: Thanks for this 😍 💯

**Review by rm3l — APPROVED** [review]

**Comment by feloy** [comment]
/override OpenShift-Integration-tests/OpenShift-Integration-tests
Unrelated flaky test
```
[FAILED] [238.740 seconds]
odo dev command tests
/go/odo_1/tests/integration/cmd_dev_test.go:31
  port-forwarding for the component
  /go/odo_1/tests/integration/cmd_dev_test.go:595
    when devfile has single endpoint
    /go/odo_1/tests/integration/cmd_dev_test.go:596
      when running odo dev
      /go/odo_1/tests/integration/cmd_dev_test.go:605
        when modifying memoryLimit for container in Devfile [BeforeEach]
        /go/odo_1/tests/integration/cmd_dev_test.go:639
          should expose the endpoint on localhost
          /go/odo_1/tests/integration/cmd_dev_test.go:655
```

**Pull request merged**